### ISSUE TITLE
lab4

### ISSUE THREAD

**Issue by dlcor** [dlcor/AI#4]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from xgboost import XGBClassifier

# Загрузка данных
df = pd.read_csv('Heart Prediction Quantum Dataset.csv')

# Первичный анализ данных
print(df.info())
print(df.describe())
print(df.head())

# Визуализация распределения целевого класса
sns.countplot(x='HeartDisease', data=df)
plt.title("Распределение классов HeartDisease")
plt.show()

# Гистограмма возраста
sns.histplot(df['Age'], bins=30, kde=True)
plt.title("Распределение возраста пациентов")
plt.xlabel("Возраст")
plt.ylabel("Частота")
plt.show()

# Анализ корреляций
plt.figure(figsize=(12, 8))
sns.heatmap(df.corr(), annot=True, cmap='coolwarm')
plt.title("Тепловая карта корреляций")
plt.show()

# Boxplot возраста по классам
sns.boxplot(x='HeartDisease', y='Age', data=df)
plt.title("Распределение возраста по классам HeartDisease")
plt.show()

# Гистограмма холестерина
sns.histplot(df['Cholesterol'], bins=30, kde=True)
plt.title("Распределение уровня холестерина")
plt.xlabel("Холестерин")
plt.ylabel("Частота")
plt.show()

# Диаграмма рассеяния
sns.scatterplot(x='QuantumPatternFeature', y='Cholesterol', hue='HeartDisease', data=df, alpha=0.6)
plt.title("Связь между математ. расчетным значением и уровнем холестерина")
plt.show()

# Разделение данных
X = df.drop('HeartDisease', axis=1).values
y = df['HeartDisease'].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

# Нормализация
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Обучение модели XGBoost
model = XGBClassifier(use_label_encoder=False, eval_metric='logloss')
model.fit(X_train, y_train)

# Предсказание
y_pred = model.predict(X_test)

# Метрики
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")
print(classification_report(y_test, y_pred))

# Матрица ошибок
conf_matrix = confusion_matrix(y_test, y_pred)
plt.figure(figsize=(6, 5))
sns.heatmap(conf_matrix, annot=True, fmt='d', cmap='Blues', xticklabels=['No Disease', 'Disease'], yticklabels=['No Disease', 'Disease'])
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.title('Матрица ошибок')
plt.show()

# Предсказание на новых данных
new_data = np.array([[58, 0, 97, 249, 89, 9.249001528]])  # Убедись, что вектор по признакам соответствует
new_data = scaler.transform(new_data)
prediction = model.predict(new_data)
print(f'Предсказание (0 - нет болезни, 1 - есть): {prediction[0]}')
